### ISSUE TITLE
OOM on GTX 1080 Ti during training with default parameters

### ISSUE THREAD

**Issue by CorentinJ** [Rayhane-mamah/Tacotron-2#266]
I'm training Tacotron2 with the default parameters, starting from a checkpoint at step 300. I assumed that a GPU of 12gb would not throw any OOM error, right? Here's the output I get after step 611:

```
Generated 64 train batches of size 32 in 6.480 sec
[2018-11-13 10:00:49.854]  Step     612 [4.192 sec/step, loss=0.95526, avg_loss=0.95303]
[2018-11-13 10:00:56.268]  Step     613 [4.223 sec/step, loss=0.92781, avg_loss=0.95255]
[2018-11-13 10:00:58.146]  Step     614 [4.207 sec/step, loss=0.94198, avg_loss=0.95225]
[2018-11-13 10:01:01.888]  Step     615 [4.205 sec/step, loss=0.94248, avg_loss=0.95203]
[2018-11-13 10:01:05.005]  Step     616 [4.191 sec/step, loss=0.93736, avg_loss=0.95197]
[2018-11-13 10:01:06.980]  Step     617 [4.171 sec/step, loss=0.97370, avg_loss=0.95214]
[2018-11-13 10:01:11.957]  Step     618 [4.202 sec/step, loss=0.94101, avg_loss=0.95182]
[2018-11-13 10:01:16.204]  Step     619 [4.200 sec/step, loss=0.93930, avg_loss=0.95166]
[2018-11-13 10:01:19.287]  Step     620 [4.214 sec/step, loss=0.93993, avg_loss=0.95124]
[2018-11-13 10:01:25.168]  Step     621 [4.237 sec/step, loss=0.93386, avg_loss=0.95094]
[2018-11-13 10:01:29.450]  Step     622 [4.230 sec/step, loss=0.94320, avg_loss=0.95070]
[2018-11-13 10:01:31.287]  Step     623 [4.196 sec/step, loss=0.93933, avg_loss=0.95060]
[2018-11-13 10:01:34.929]  Step     624 [4.203 sec/step, loss=0.93553, avg_loss=0.95018]
[2018-11-13 10:01:40.721]  Step     625 [4.199 sec/step, loss=0.92288, avg_loss=0.95008]
[2018-11-13 10:01:45.923]  Step     626 [4.192 sec/step, loss=0.92520, avg_loss=0.94993]
[2018-11-13 10:01:51.599]  Step     627 [4.191 sec/step, loss=0.91564, avg_loss=0.94957]
[2018-11-13 10:01:55.252]  Step     628 [4.185 sec/step, loss=0.92450, avg_loss=0.94923]
[2018-11-13 10:02:00.769]  Step     629 [4.215 sec/step, loss=0.92467, avg_loss=0.94888]
[2018-11-13 10:02:06.378]  Step     630 [4.224 sec/step, loss=0.92030, avg_loss=0.94857]
[2018-11-13 10:02:10.999]  Step     631 [4.257 sec/step, loss=0.91963, avg_loss=0.94841]
[2018-11-13 10:02:12.431]  Step     632 [4.221 sec/step, loss=0.92896, avg_loss=0.94814]
[2018-11-13 10:02:14.053]  Step     633 [4.176 sec/step, loss=0.92970, avg_loss=0.94790]
[2018-11-13 10:02:16.585]  Step     634 [4.149 sec/step, loss=0.95214, avg_loss=0.94785]
[2018-11-13 10:02:18.234]  Step     635 [4.118 sec/step, loss=0.95411, avg_loss=0.94778]
[2018-11-13 10:02:23.522]  Step     636 [4.141 sec/step, loss=0.92869, avg_loss=0.94734]
[2018-11-13 10:02:29.127]  Step     637 [4.161 sec/step, loss=0.92654, avg_loss=0.94687]
[2018-11-13 10:02:34.817]  Step     638 [4.197 sec/step, loss=0.92537, avg_loss=0.94628]
[2018-11-13 10:02:35.894]  Step     639 [4.156 sec/step, loss=0.89254, avg_loss=0.94560]
[2018-11-13 10:02:39.602]  Step     640 [4.170 sec/step, loss=0.96091, avg_loss=0.94535]
[2018-11-13 10:02:43.334]  Step     641 [4.151 sec/step, loss=0.93454, avg_loss=0.94516]
[2018-11-13 10:02:48.011]  Step     642 [4.138 sec/step, loss=0.95098, avg_loss=0.94522]
[2018-11-13 10:02:51.853]  Step     643 [4.137 sec/step, loss=0.93831, avg_loss=0.94498]
[2018-11-13 10:02:57.366]  Step     644 [4.156 sec/step, loss=0.92009, avg_loss=0.94452]
[2018-11-13 10:03:03.230]  Step     645 [4.184 sec/step, loss=0.92807, avg_loss=0.94415]
[2018-11-13 10:03:09.186]  Step     646 [4.231 sec/step, loss=0.91936, avg_loss=0.94423]
[2018-11-13 10:03:14.286]  Step     647 [4.231 sec/step, loss=0.92024, avg_loss=0.94381]
[2018-11-13 10:03:18.608]  Step     648 [4.210 sec/step, loss=0.92471, avg_loss=0.94353]
[2018-11-13 10:03:22.622]  Step     649 [4.204 sec/step, loss=0.93032, avg_loss=0.94318]
[2018-11-13 10:03:25.534]  Step     650 [4.178 sec/step, loss=0.95527, avg_loss=0.94328]
[2018-11-13 10:03:27.471]  Step     651 [4.143 sec/step, loss=0.95260, avg_loss=0.94330]
[2018-11-13 10:03:31.652]  Step     652 [4.141 sec/step, loss=0.92412, avg_loss=0.94278]
[2018-11-13 10:03:35.338]  Step     653 [4.146 sec/step, loss=0.94663, avg_loss=0.94265]
[2018-11-13 10:03:40.132]  Step     654 [4.160 sec/step, loss=0.92890, avg_loss=0.94221]
[2018-11-13 10:03:45.060]  Step     655 [4.166 sec/step, loss=0.91632, avg_loss=0.94191]
[2018-11-13 10:03:48.538]  Step     656 [4.151 sec/step, loss=0.92222, avg_loss=0.94160]
[2018-11-13 10:03:52.726]  Step     657 [4.147 sec/step, loss=0.93296, avg_loss=0.94129]
[2018-11-13 10:03:57.091]  Step     658 [4.153 sec/step, loss=0.92029, avg_loss=0.94101]
[2018-11-13 10:04:02.751]  Step     659 [4.186 sec/step, loss=0.92316, avg_loss=0.94033]
[2018-11-13 10:04:07.996]  Step     660 [4.196 sec/step, loss=0.92487, avg_loss=0.94008]
[2018-11-13 10:04:12.498]  Step     661 [4.182 sec/step, loss=0.92278, avg_loss=0.93984]
[2018-11-13 10:04:14.968]  Step     662 [4.146 sec/step, loss=0.91624, avg_loss=0.93979]
[2018-11-13 10:04:19.177]  Step     663 [4.138 sec/step, loss=0.92684, avg_loss=0.93957]
[2018-11-13 10:04:32.492]  Exiting due to exception: OOM when allocating tensor with shape[32,799,185] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
         [[node Tacotron_model/loss/strided_slice (defined at /home/cjemine/tacotron2/tacotron/models/tacotron.py:318)  = StridedSlice[Index=DT_INT32, T=DT_FLOAT, begin_mask=3, ellipsis_mask=0, end_mask=3, new_axis_mask=0, shrink_axis_mask=0, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Tacotron_model/loss/Abs, Tacotron_model/inference/TacoTrainingHelper/strided_slice/stack, Tacotron_model/loss/strided_slice/stack_1, Tacotron_model/optimizer_1/gradients/Tacotron_model/loss/mean_squared_error/Sum_grad/Reshape/shape)]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

         [[{{node Tacotron_model/clip_by_global_norm/mul_99/_953}} = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_12474_Tacotron_model/clip_by_global_norm/mul_99", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.


Caused by op 'Tacotron_model/loss/strided_slice', defined at:
  File "/home/cjemine/.pycharm_helpers/pydev/pydevd.py", line 1664, in <module>
    main()
  File "/home/cjemine/.pycharm_helpers/pydev/pydevd.py", line 1658, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/home/cjemine/.pycharm_helpers/pydev/pydevd.py", line 1068, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/home/cjemine/.pycharm_helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/cjemine/tacotron2/train.py", line 138, in <module>
    main()
  File "/home/cjemine/tacotron2/train.py", line 132, in main
    train(args, log_dir, hparams)
  File "/home/cjemine/tacotron2/train.py", line 52, in train
    checkpoint = tacotron_train(args, log_dir, hparams)
  File "/home/cjemine/tacotron2/tacotron/train.py", line 376, in tacotron_train
    return train(log_dir, args, hparams)
  File "/home/cjemine/tacotron2/tacotron/train.py", line 156, in train
    model, stats = model_train_mode(args, feeder, hparams, global_step)
  File "/home/cjemine/tacotron2/tacotron/train.py", line 92, in model_train_mode
    model.add_loss()
  File "/home/cjemine/tacotron2/tacotron/models/tacotron.py", line 318, in add_loss
    linear_loss = 0.5 * tf.reduce_mean(l1) + 0.5 * tf.reduce_mean(l1[:,:,0:n_priority_freq])
  File "/home/cjemine/miniconda3/lib/python3.6/site-packages/tensorflow/python/ops/array_ops.py", line 525, in _slice_helper
    name=name)
  File "/home/cjemine/miniconda3/lib/python3.6/site-packages/tensorflow/python/ops/array_ops.py", line 691, in strided_slice
    shrink_axis_mask=shrink_axis_mask)
  File "/home/cjemine/miniconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 8522, in strided_slice
    name=name)
  File "/home/cjemine/miniconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/cjemine/miniconda3/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/cjemine/miniconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/home/cjemine/miniconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[32,799,185] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
         [[node Tacotron_model/loss/strided_slice (defined at /home/cjemine/tacotron2/tacotron/models/tacotron.py:318)  = StridedSlice[Index=DT_INT32, T=DT_FLOAT, begin_mask=3, ellipsis_mask=0, end_mask=3, new_axis_mask=0, shrink_axis_mask=0, _device="/job:localhost/replica:0/task:0/device:GPU:0"](Tacotron_model/loss/Abs, Tacotron_model/inference/TacoTrainingHelper/strided_slice/stack, Tacotron_model/loss/strided_slice/stack_1, Tacotron_model/optimizer_1/gradients/Tacotron_model/loss/mean_squared_error/Sum_grad/Reshape/shape)]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.

         [[{{node Tacotron_model/clip_by_global_norm/mul_99/_953}} = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_12474_Tacotron_model/clip_by_global_norm/mul_99", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.
```

Also, why do I have a checkpoint at 300 when the checkpoint interval is 5000 by default?

**Comment by begeekmyfriend** [comment]
If you have single GPU, try setting
```python
max_mel_frames = 900,
outputs_per_step = 2,
```

**Comment by CorentinJ** [comment]
In hparams.py it says: "Never use lower batch size than 32 on a single GPU!". Does anyone know the reason why?

I'm using @begeekmyfriend's suggestion as a workaround for now but I'm still aiming for high audio quality in the long term.

**Comment by Rayhane-mamah** [comment]
Use max_mel_frames=800 and outputs_per_step=1 should be fine.
We don't recommend using batch size lower than 32 because batchnorm
requires representative batches for efficient training.

**Comment by CorentinJ** [comment]
800 max frames works, thanks

**Comment by begeekmyfriend** [comment]
Try outputs_per_step = 2

**Comment by statcom** [comment]
Just for reference, GTX 1080 seemed to work fine with max_mel_frames=500 and outputs_per_step=1
